src/crimson/osd/osd_operation.h

```diff
@@ -0,0 +1,149 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <vector>
+#include <set>
+#include <boost/intrusive/list.hpp>
+#include <boost/intrusive_ptr.hpp>
+#include <boost/smart_ptr/intrusive_ref_counter.hpp>
+
+#include "common/Formatter.h"
+
+namespace ceph::osd {
+
+enum class OperationTypeCode {
+  client_request = 0,
+  peering_event,
+  compound_peering_request,
+  last_op
+};
+
+static constexpr const char* const OP_NAMES[] = {
+  "client_write",
+  "peering_event",
+  "compound_peering_request"
+};
+
+class OperationRegistry;
+
+using registry_hook_t = boost::intrusive::list_member_hook<
+  boost::intrusive::link_mode<boost::intrusive::auto_unlink>>;
+
+class Blocker;
+class Operation : public boost::intrusive_ref_counter<
+  Operation, boost::thread_unsafe_counter> {
+  friend class OperationRegistry;
+  registry_hook_t registry_hook;
+
+  std::set<Blocker*> blockers;
```

Comment:
Do you need the ordering of the 'set' here? would a linear search of a (supposedly very short, isn't it?) std::vector, or an unordered-set be faster?